File: practice_problems/arrays/ruby/xor_diff.rb
# Objective: Given two arays, return a new array containing elements that are in
# either one or the other but not both
# NOTE: This isn't a bit manipulation problem, despite the 'xor' in the name

def xor_diff(arr1, arr2)
  hashed = Hash.new(0)

  arr1.each do |el|
    hashed[el] = 'first'
  end

  arr2.each do |el|
    hashed[el] = hashed[el] == 'first' ?  'both' : 'second'
  end

  hashed.delete_if {|key,value| value == 'both'}
  hashed.keys
end

p xor_diff([1,2], [2,3])

File: practice_problems/arrays/python/xor_diff.py
def xor_diff(arr1, arr2):
    dict = {}

    for el1 in arr1:
        dict[el1] = 'a'

    for el2 in arr2:
        check = el2 in dict
        dict[el2] = 'ab' if check else 'b'

    items = [key for key, val in dict.items() if val == 'ab']

    for dupe in items:
        del dict[dupe]

    return dict.keys()

print(xor_diff([1,2,3], [1,5,'a']))
